uaclient/api/u/pro/detach/v1.py

```diff
@@ -0,0 +1,102 @@
+from typing import List
+
+from uaclient import daemon, entitlements, exceptions, messages, timer, util
+from uaclient.api.api import APIEndpoint
+from uaclient.api.data_types import AdditionalInfo, ErrorWarningObject
+from uaclient.api.u.pro.security.status.reboot_required.v1 import (
+    _reboot_required,
+)
+from uaclient.api.u.pro.status.is_attached.v1 import _is_attached
+from uaclient.config import UAConfig
+from uaclient.data_types import (
+    BoolDataValue,
+    DataObject,
+    Field,
+    StringDataValue,
+    data_list,
+)
+from uaclient.files import state_files
+from uaclient.timer.update_messaging import update_motd_messages
+
+
+class DetachResult(DataObject, AdditionalInfo):
+    fields = [
+        Field("disabled", data_list(StringDataValue)),
+        Field("reboot_required", BoolDataValue),
+    ]
+
+    def __init__(self, disabled: List[str], reboot_required: bool):
+        self.disabled = disabled
+        self.reboot_required = reboot_required
+
+
+def detach() -> DetachResult:
+    return _detach(UAConfig())
+
+
+def _detach(cfg: UAConfig) -> DetachResult:
+    if not util.we_are_currently_root():
+        raise exceptions.NonRootUserError
+
+    if not _is_attached(cfg).is_attached:
+        return DetachResult(
+            disabled=[],
+            reboot_required=False,
+        )
+
+    disabled = []
+    warnings = []  # type: List[ErrorWarningObject]
+    for ent_name in entitlements.entitlements_disable_order(cfg):
+        try:
+            ent_cls = entitlements.entitlement_factory(cfg=cfg, name=ent_name)
+        except exceptions.EntitlementNotFoundError:
+            continue
+
+        ent = ent_cls(cfg=cfg, assume_yes=True)
+        # For detach, we should not consider that a service
+        # cannot be disabled because of dependent services,
+        # since we are going to disable all of them anyway
+        can_disable, _ = ent.can_disable(ignore_dependent_services=True)
+        if can_disable:
```

Comment:
Do we need this check? `disable` will call it and disable dependent services with `assume_yes=True`

Reply:
The rationale here is to avoid having warnings for services that should not be disabled anyway, as we are looping on every service we have.

I can remove that check, if we instead loop into only services returned by `_enabled_services`. But let me know what you think ?

Reply:
Ohhh now I understand - I do think the intent would be more clear if we looped over `enabled_services` instead, but this is fine too - up to you.